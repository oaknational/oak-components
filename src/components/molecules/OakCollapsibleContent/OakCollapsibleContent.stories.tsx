import React, { useState } from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakCollapsibleContent } from "./OakCollapsibleContent";

import { OakTertiaryButton } from "@/components/molecules/OakTertiaryButton";
import { OakP } from "@/components/atoms";
import { sizeArgTypes } from "@/storybook-helpers/sizeStyleHelpers";

const meta: Meta<typeof OakCollapsibleContent> = {
  component: OakCollapsibleContent,
  tags: ["autodocs"],
  title: "components/molecules/OakCollapsibleContent",
  argTypes: {
    ...sizeArgTypes,
  },
  parameters: {
    controls: {
      include: ["isOpen", ...Object.keys(sizeArgTypes)],
    },
  },
  args: {
    isOpen: true,
    $maxHeight: "all-spacing-19",
    children: (
      <>
        <OakP $mb="space-between-m">
          Hello, I'm Mr. Norris, and welcome to the first lesson in this topic
          on forces for year seven.
        </OakP>
        <OakP $mb="space-between-m">
          This lesson is all about what forces do. This is part of the wider big
        </OakP>
        <OakP $mb="space-between-m">
          question of how forces make things happen in the universe. Everything
          that
        </OakP>
        <OakP $mb="space-between-m">
          happens in the universe at the deepest level, any change happens,
          because
        </OakP>
        <OakP $mb="space-between-m">
          a force has acted. And that's why this is such a great, such an
          important
        </OakP>
        <OakP>
          topic to do at the start of year seven. So let's get started.
        </OakP>
      </>
    ),
  },
};
export default meta;

type Story = StoryObj<typeof OakCollapsibleContent>;

export const Default: Story = {
  render: (args) => <OakCollapsibleContent {...args} />,
};

export const WithControl: Story = {
  render: (args) => {
    const [isOpen, setIsOpen] = useState(true);
    return (
      <>
        <OakTertiaryButton
          onClick={() => setIsOpen(!isOpen)}
          iconName={isOpen ? "chevron-up" : "chevron-down"}
          $mb="space-between-m"
          aria-expanded={isOpen}
          aria-controls="collapsible-content"
        >
          {isOpen ? "Close" : "Open"}
        </OakTertiaryButton>
        <OakCollapsibleContent
          id="collapsible-content"
          {...args}
          isOpen={isOpen}
        />
      </>
    );
  },
};
