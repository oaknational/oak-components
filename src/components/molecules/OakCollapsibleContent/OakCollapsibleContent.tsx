import React, { ComponentProps } from "react";
import styled from "styled-components";

import { OakBox, OakFlex } from "@/components/atoms";
import { parseColor } from "@/styles/helpers/parseColor";
import { parseSpacing } from "@/styles/helpers/parseSpacing";
import { parseBorderRadius } from "@/styles/helpers/parseBorderRadius";

export type OakCollapsibleContentProps = ComponentProps<typeof OakFlex> & {
  /**
   * Whether the content is displayed or not
   */
  isOpen: boolean;
};

const StyledScrollBox = styled(OakBox)`
  &::-webkit-scrollbar-thumb {
    border-radius: ${parseBorderRadius("border-radius-m")};
    background-color: ${parseColor("grey50")};
  }
  &::-webkit-scrollbar-track {
    border-radius: ${parseBorderRadius("border-radius-m")};
    background-color: ${parseColor("white")};
  }
  &::-webkit-scrollbar {
    width: ${parseSpacing("spacing-12")};
    height: ${parseSpacing("spacing-12")};
    border-radius: ${parseBorderRadius("border-radius-m")};
  }
`;

/**
 * A filled scrollable box that can be hidden with the `isOpen` prop.
 * Useful for hiding content that is not immediately relevant
 */
export const OakCollapsibleContent = ({
  isOpen,
  children,
  id,
  ...rest
}: OakCollapsibleContentProps) => {
  return (
    <OakBox $display={isOpen ? "block" : "none"} id={id}>
      <OakFlex
        $background="bg-neutral-stronger"
        $ph="spacing-16"
        $pv="spacing-24"
        $borderRadius="border-radius-m"
        {...rest}
      >
        <StyledScrollBox
          $overflow="auto"
          $maxHeight="100%"
          $width="100%"
          $pr="spacing-16"
        >
          {children}
        </StyledScrollBox>
      </OakFlex>
    </OakBox>
  );
};
