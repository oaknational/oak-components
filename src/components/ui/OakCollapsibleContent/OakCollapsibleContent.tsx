import React, { ComponentProps } from "react";
import styled from "styled-components";

import { OakBox, OakFlex } from "@/components/base";
import { parseColor } from "@/styles/helpers/parseColor";
import { parseSpacing } from "@/styles/helpers/parseSpacing";
import { parseBorderRadius } from "@/styles/helpers/parseBorderRadius";

export type OakCollapsibleContentProps = ComponentProps<typeof OakFlex> & {
  /**
   * Whether the content is displayed or not
   */
  isOpen: boolean;
};

const StyledScrollBox = styled(OakBox)`
  &::-webkit-scrollbar-thumb {
    border-radius: ${parseBorderRadius("border-radius-m")};
    background-color: ${parseColor("grey50")};
  }
  &::-webkit-scrollbar-track {
    border-radius: ${parseBorderRadius("border-radius-m")};
    background-color: ${parseColor("white")};
  }
  &::-webkit-scrollbar {
    width: ${parseSpacing("all-spacing-3")};
    height: ${parseSpacing("all-spacing-3")};
    border-radius: ${parseBorderRadius("border-radius-m")};
  }
`;

/**
 * A filled scrollable box that can be hidden with the `isOpen` prop.
 * Useful for hiding content that is not immediately relevant
 */
export const OakCollapsibleContent = ({
  isOpen,
  children,
  id,
  ...rest
}: OakCollapsibleContentProps) => {
  return (
    <OakBox $display={isOpen ? "block" : "none"} id={id}>
      <OakFlex
        $background="bg-neutral-stronger"
        $ph="inner-padding-m"
        $pv="inner-padding-xl"
        $borderRadius="border-radius-m"
        {...rest}
      >
        <StyledScrollBox
          $overflow="auto"
          $maxHeight="100%"
          $width="100%"
          $pr="inner-padding-m"
        >
          {children}
        </StyledScrollBox>
      </OakFlex>
    </OakBox>
  );
};
