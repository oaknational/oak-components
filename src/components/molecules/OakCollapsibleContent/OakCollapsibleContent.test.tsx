import React from "react";
import "@testing-library/jest-dom";

import { OakCollapsibleContent } from "./OakCollapsibleContent";

import renderWithTheme from "@/test-helpers/renderWithTheme";

describe(OakCollapsibleContent, () => {
  it("matches snapshot", () => {
    const { container } = renderWithTheme(
      <OakCollapsibleContent isOpen>Content goes here</OakCollapsibleContent>,
    );

    expect(container).toMatchSnapshot();
  });
});
