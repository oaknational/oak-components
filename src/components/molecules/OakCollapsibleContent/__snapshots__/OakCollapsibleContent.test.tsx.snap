// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakCollapsibleContent matches snapshot 1`] = `
.c0 {
  display: block;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c1 {
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 1.5rem;
  padding-bottom: 1.5rem;
  background: #e4e4e4;
  border-radius: 0.375rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c3 {
  overflow: auto;
  width: 100%;
  max-height: 100%;
  padding-right: 1rem;
  font-family: --var(google-font),Lexend,sans-serif;
  overflow: auto;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4::-webkit-scrollbar-thumb {
  border-radius: 0.375rem;
  background-color: #808080;
}

.c4::-webkit-scrollbar-track {
  border-radius: 0.375rem;
  background-color: #ffffff;
}

.c4::-webkit-scrollbar {
  width: 0.75rem;
  height: 0.75rem;
  border-radius: 0.375rem;
}

<div
  className="c0"
>
  <div
    className="c1 c2"
  >
    <div
      className="c3 c4"
    >
      Content goes here
    </div>
  </div>
</div>
`;
